bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java

```diff
@@ -0,0 +1,288 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.broadlink.handler;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.broadlink.BroadlinkBindingConstants;
+import org.openhab.binding.broadlink.config.BroadlinkDeviceConfiguration;
+import org.openhab.binding.broadlink.internal.BroadlinkProtocol;
+import org.openhab.binding.broadlink.internal.NetworkUtils;
+import org.openhab.binding.broadlink.internal.Utils;
+import org.openhab.binding.broadlink.internal.discovery.DeviceRediscoveryAgent;
+import org.openhab.binding.broadlink.internal.discovery.DeviceRediscoveryListener;
+import org.openhab.binding.broadlink.internal.socket.NetworkTrafficObserver;
+import org.openhab.binding.broadlink.internal.socket.RetryableSocket;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.util.HexUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract superclass of all supported Broadlink devices.
+ *
+ * @author John Marshall/Cato Sognen - Initial contribution
+ */
+@NonNullByDefault
+public abstract class BroadlinkBaseThingHandler extends BaseThingHandler implements DeviceRediscoveryListener {
+
+    public final Logger logger = LoggerFactory.getLogger(getClass());
+    private static final String INITIAL_DEVICE_ID = "00000000";
+
+    protected BroadlinkDeviceConfiguration thingConfig;
+
+    private @Nullable RetryableSocket socket;
+
+    // Can be injected for test purposes
+    private @Nullable NetworkTrafficObserver networkTrafficObserver;
+
+    private int count;
+    private @Nullable ScheduledFuture<?> refreshHandle;
+    private boolean authenticated = false;
+    // These get handed to us by the device after successful authentication:
+    private byte[] deviceId;
+    private byte[] deviceKey;
+
+    public BroadlinkBaseThingHandler(Thing thing, Logger logger) {
+        super(thing);
+        this.thingConfig = getConfigAs(BroadlinkDeviceConfiguration.class);
```

Comment:
I could be wrong but if you leave it here the config may not get updated when someone edits a config field AFTER the thing is already created and saved. When someone edits I believe the handler does not get re-created as the dispose() is called and then the initialize(). If you move this line down into the initialize() method then it will ensure the latest changes are in the thingConfig instance.

Reply:
That makes a lot of sense - for a long time people have complained about changes not showing up. Many thanks!